midproject

qianyang1997 opened the pull request:
Hey LC,

Feel free to run the commands in the Midproject Checkpoint section of the [Readme file](https://github.com/MSIA/2021-msia423-yang-qiana-project/blob/develop/README.md) and let me know if it works on a Mac!

In the most recent commits, I:
1) created a [script](https://github.com/MSIA/2021-msia423-yang-qiana-project/blob/develop/src/ingest.py) to download raw data and upload it to S3 without saving it locally;
2) created a [script](https://github.com/MSIA/2021-msia423-yang-qiana-project/blob/develop/src/create_db.py) to build a database schema either locally or in AWS RDS. The schema involves two (empty) tables named 'survey' and 'metadata' respectively;
3) created shell scripts (see Readme) to run the above Python scripts via a Docker virtual environment;
4) created a [flaskconfig.py file](https://github.com/MSIA/2021-msia423-yang-qiana-project/blob/develop/config/flaskconfig.py) to store variables used in ingest.py and create_db.py for cleanness of code;
5) added command line arguments in run.py in the root directory;
5) adjusted requirements.txt, .gitignore, README.md, and other helper files accordingly.

When you examine my code, pay attention to:
1) logging (set different logging configuration levels in config.py and see if my logs are clear and helpful)
2) optional command line arguments (try different combinations of custom inputs and see if anything breaks)
3) coding style and best practices (point out any redundancy in code)
4) output (e.g. download the raw data from S3 locally and see if the csv file makes sense)

Hope you enjoy being my QA!

lcgenereux13 commented:
Hey Qiana, absolute amazing job!

First, everything runs smoothly, a big congrats. 

Below are details on the various steps I took to ensure that the code runs as expected (see section 1). I then addressed some of the QA requests you made below (see section 2). Please also see the attached pdf with more details (including screenshots). I hope this is useful!


# SECTION 1: RUNNING CODE


## STEP 1: Ensure that docker image works: SUCCESS

I successfully created the docker image you built.



## STEP 2: Ensure that data can be downloaded to S3: SUCCESS

I ran three separate tests here:

* TEST 1: NO codebook or datapath specified. SUCCESS
    * Confirms that when codebook and data paths are un-specified, your defaults are used for codebook and data path

* TEST 2: Codebook or datapath both specified. SUCCESS
    * All files were downloaded to the desired destination

* TEST 3: Only one of code_book or data path specified. SUCCESS
    * Codebook goes in “raw” as expected while data goes in “path_test” as specified


## STEP 3A: Entering mysql with my QA credentials: SUCCESS

* My credentials work and I am only able to view the msia423_db, as expected. 
* I have no privileges on your other databases


## STEP 3B: Recreating the MSIA423_db: SUCCESS

## STEP 4: Ensuring that tables have been created in db: SUCCESS

* Point of feedback
    * I see that the “tracks” (which was not created through your code) is still there. 
    * Perhaps this was by design, but I believe this could be removed in the next iteration of your project


## Step 5: Checking that tables match with schemas that were specified in your code: SUCCESS

* Point of feedback:
    * In creating the ‘country’ field in the metadata table and the ‘question’ field in the survey table, you specified a variable length string of maximum length 10. 
    * To avoid cases where the question or country cannot fit in ten characters, I would consider lengthening these strings



# SECTION 2: COVERING YOUR QA REQUESTS

Beyond ensuring that the code runs smoothly in docker (including downloading of data, database creation, etc.), I moved on to the other areas of focus which you suggested


##  1. Seting different logging configuration levels
* I have changed the logging configuration. Your info logs were informative and insightful as I ran your commands

## 2. Testing different command line combinations 
 * Your parsers and default settings work beautifully. As mentionned above, I was able to download data with no specified paths, with one specified path for the code book / anoter for the data and finally with a specified path for both the codebook and the data

## 3. Coding style and best practice:
* Non required import:
    * ‘Import argparse’ in src.ingest.py is not required. While commands gathered through the parser are used to run the download() function in src.ingest, the parser is called in the run.py file (where the download() function is called from); therefore it only needs to be imported in the run.py file. ‘Import argparse ‘was also specified in create_db() and is not required for the same reason.
* Presence of music related table and .py file
    * Your repo still includes the add songs.py file (while it is all “commented”, I think it could be removed).
    * As mentionned in section 1, your “tracks” table is still in the db as well, and I believe it won’t be used for your project.
* Possible removal of unused text:
    * You stil have a long dosctring below your config file (I believe it includes old code). This could be deleted for stylistic purposes
* Opportunities to provide more info to other users (solely based on personal preferences) 
    * I would specify what exactly “codebook” refers to in the readme file. I initially thought this refered to python code before (silly I know, but I think future users would benefit from knowing that it is the questions bank for your survey response data)
    * I would add a comment on line 11 of run.py to specify that the following sub-parser relates to the downloading the data. I find this approach of using a parser and subparcer to download data and create a db from just 1 .py file (run.py) quite nice, good job!

## 4. Downloading and inspecting data:
* I downloaded data.csv from S3 to my machine and tested that the output is what we want. 
    * I see 169 colums: 163 of which report (0 to 5 scaled) answers to questions (identified only by a letter-number combination). The last 6 columns are combination of demographic and respondent source data. There are 49,160 observations in the dataset I downloaded.
    * I also dowloaded codebook.txt. It maps column identifiers to full survey questions and is formatted in html blocks, this should make parsing easy.
    * Is this what you expected?

These are my observations! Again, amazing job, everything ran very smoothly and was clearly laid out in your “read me” file! I will be online today and Friday for any other possible reviews :)

Cheers!

LC

lcgenereux13 commented:
[QA on Qiana project (midterm).pdf](https://github.com/MSIA/2021-msia423-yang-qiana-project/files/6513281/QA.on.Qiana.project.midterm.pdf)

qianyang1997 commented:
Hi LC,

Awesome review - thank you!

1) Thanks for catching 'import argparse' in my src modules. I have deleted them.
2) I have added a brief note in README specifying the content of 'codebook.txt'.
3) I am in the process of incorporating all code from the template repo to my scripts. By EOD, I will try to have as much old code deleted as possible.
4) Thanks for examining the content of my csv file and codebook. What you described is exactly what I intended to upload. Thanks again for checking!

Will post an update soon!

Qiana
